src/storage/exec/ScanNode.h

```diff
@@ -230,28 +259,45 @@ class ScanEdgePropNode : public QueryNode<Cursor> {
     nebula::cpp2::ErrorCode ret = nebula::cpp2::ErrorCode::SUCCEEDED;
     for (auto& edgeNode : edgeNodes_) {
       ret = edgeNode->collectEdgePropsIfValid(
-          [&row](const std::vector<PropContext>* props) -> nebula::cpp2::ErrorCode {
+          [&row, edgeNode = edgeNode.get(), this](
+              const std::vector<PropContext>* props) -> nebula::cpp2::ErrorCode {
             for (const auto& prop : *props) {
               if (prop.returned_) {
                 row.emplace_back(Value());
               }
+              if (prop.filtered_ && expCtx_ != nullptr) {
+                expCtx_->setEdgeProp(edgeNode->getEdgeName(), prop.name_, Value());
+              }
             }
             return nebula::cpp2::ErrorCode::SUCCEEDED;
           },
-          [&row, vIdLen, isIntId](
+          [&row, vIdLen, isIntId, edgeNode = edgeNode.get(), this](
               folly::StringPiece key,
               RowReader* reader,
               const std::vector<PropContext>* props) -> nebula::cpp2::ErrorCode {
-            if (!QueryUtils::collectEdgeProps(key, vIdLen, isIntId, reader, props, row).ok()) {
-              return nebula::cpp2::ErrorCode::E_EDGE_PROP_NOT_FOUND;
+            for (const auto& prop : *props) {
+              if (prop.returned_ || (prop.filtered_ && expCtx_ != nullptr)) {
+                auto value = QueryUtils::readEdgeProp(key, vIdLen, isIntId, reader, prop);
+                if (!value.ok()) {
+                  return nebula::cpp2::ErrorCode::E_EDGE_PROP_NOT_FOUND;
+                }
+                if (prop.filtered_ && expCtx_ != nullptr) {
+                  expCtx_->setEdgeProp(edgeNode->getEdgeName(), prop.name_, value.value());
+                }
+                if (prop.returned_) {
+                  VLOG(2) << "Collect prop " << prop.name_;
+                  row.emplace_back(std::move(value).value());
+                }
+              }
             }
             return nebula::cpp2::ErrorCode::SUCCEEDED;
           });
       if (ret != nebula::cpp2::ErrorCode::SUCCEEDED) {
         break;
       }
     }
-    if (ret == nebula::cpp2::ErrorCode::SUCCEEDED) {
+    if (ret == nebula::cpp2::ErrorCode::SUCCEEDED &&
+        (filter_ == nullptr || vTrue(filter_->eval(*expCtx_)))) {
```

Comment:
Maybe we need to clear every value in `expCtx_`? Otherwise when you move on to next key-value, the eval result could be wrong?

Reply:
Good catch.Fixed.